### PR TITLE
create-or-update-issue: add action

### PR DESCRIPTION
**Pull request by ZhongRuoyu**
Related: https://github.com/Homebrew/homebrew-core/pull/190657#issuecomment-2353378549

### CODE REVIEW

File: create-issue/action.yml [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+  color: blue
+inputs:
+  token:
+    description: GitHub token
+    required: false
+    default: ${{github.token}}
+  repository:
+    description: Repository to send the issue to
+    required: false
+    default: ${{github.repository}}
```
Comment (SMillerDev) [inline create-issue/action.yml:15]: Would be nice to have a linter for this
```suggestion
    default: ${{ github.token }}
  repository:
    description: Repository to send the issue to
    required: false
    default: ${{ github.repository }}
```

**Review by SMillerDev — APPROVED** [review]

**Review by SMillerDev — APPROVED** [review]

**Review by MikeMcQuaid — COMMENTED** [review]
Looks good! I think it might be nice to make this a little higher level such that it will also autoclose an issue on success and open a new issue on failure. Maybe even post more comments on repeated failures? Thoughts?

**Comment by ZhongRuoyu** [comment]
Sorry, have been a little bit busy this week. I'll get back to this soon.

> I think it might be nice to make this a little higher level such that it will also autoclose an issue on success and open a new issue on failure. Maybe even post more comments on repeated failures? Thoughts?

How about this: we turn this into an action that's able to perform following according to an input `${{ steps.some-previous-step.conclusion }}`:

- `failure`: looks for an issue with the same title, and post a new comment under it if found, or open a new one otherwise.
- `success`: closes an issue if there exists one with the title.

With usage like:

```yaml
- uses: Homebrew/actions/create-issue@master
  with:
    token: ${{ github.token }}
    repository: ${{ github.repository }}
    update-existing: true
    close-existing: true
    dependent-step-conclusion: ${{ steps.deploy.conclusion }} 
    title: Deployment failed
    body: Issue body (or comment to post if issue already exists)
    labels: label1,label2
    assignees: user1,user2
```

**Comment by MikeMcQuaid** [comment]
@ZhongRuoyu This seems great!

File: create-issue/action.yml [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+      if such an issue is found and `dependent-step-conclusion` is `failure`
+    required: false
+    default: "false"
+  close-existing:
+    description: >
+      Whether to close an existing issue with the same title as completed,
+      if such an issue is found and `dependent-step-conclusion` is `success`
+    required: false
+    default: "false"
+runs:
```
Comment (Bo98) [inline create-issue/action.yml:45]: I forgot to do this for `create-pull-request`, but you can add `outputs` here too: https://github.com/Homebrew/actions/blob/master/setup-homebrew/action.yml#L36.

File: create-issue/action.yml [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+      Whether to post `body` as a comment under the issue with the same title,
+      if such an issue is found and `dependent-step-conclusion` is `failure`
+    required: false
+    default: "false"
+  close-existing:
+    description: >
+      Whether to close an existing issue with the same title as completed,
+      if such an issue is found and `dependent-step-conclusion` is `success`
+    required: false
+    default: "false"
```
Comment (Bo98) [inline create-issue/action.yml:44]: Sounds like this coud be folded to:

```
update-existing: ${{ steps.<step-id>.conclusion == 'failure' }}
close-existing: ${{ steps.<step-id>.conclusion == 'success' }}
```

And be able to work much more generically without needing to be specifically tied to a `conclusion`?

File: create-issue/main.test.mjs [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+    "Content-Type": "application/json",
+  }).reply(200, {
+    number: issueNumber,
+  });
+
+  await loadMain();
+});
+
+test("create-issue-advanced", async () => {
+  const mockPool = githubMockPool();
```
Comment (Bo98) [inline create-issue/main.test.mjs:43]: You can share some code by having a `describe` with nested `it`. See `post-comment` or `api-commit-and-push` as an example.

File: create-issue/action.yml (hunk 3 above)
Comment (ZhongRuoyu) [reply create-issue/action.yml:44]: Sounds good!

File: create-or-update-issue/main.mjs [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+    const title = core.getInput("title", { required: true });
+    const body = core.getInput("body", { required: true });
+
+    const labelsInput = core.getInput("labels");
+    const labels = labelsInput ? labelsInput.split(",") : [];
+    const assigneesInput = core.getInput("assignees");
+    const assignees = assigneesInput ? assigneesInput.split(",") : [];
+
+    const updateExisting = core.getInput("update-existing") === "true";
+    const closeExisting = core.getInput("close-existing") === "true";
```
Comment (Bo98) [inline create-or-update-issue/main.mjs:19]:
```suggestion
    const updateExisting = core.getBooleanInput("update-existing");
    const closeExisting = core.getBooleanInput("close-existing");
```

File: create-or-update-issue/README.md [hunk 6]
```diff
@@ -0,0 +1,24 @@
+# Create Or Update Issue
+
+An action to create or update an issue in a repository.
+It supports posting a comment under an existing issue with the same title or
+closing it based on the outcome of a previous step.
+
+## Usage
+
+```yaml
+- uses: Homebrew/actions/create-issue@master
```
Comment (Bo98) [inline create-or-update-issue/README.md:10]:
```suggestion
- uses: Homebrew/actions/create-or-update-issue@master
```

**Review by Bo98 — APPROVED** [review]
Looks great!

**Review by MikeMcQuaid — APPROVED** [review]
Looks great, thanks @ZhongRuoyu! After merging: could you update formulae.brew.sh and homebrew/core to use this action and, if you're feeling really generous, add to the Homebrew/brew release workflow that uploads packages 🙇🏻

**Comment by ZhongRuoyu** [comment]
Sure! I'll first do formulae.brew.sh as that's run most frequently. If that works well I'll roll this out to more places.

**Pull request merged**